### PR TITLE
File storage service (KML)

### PR DESCRIPTION
**Pull request by procrastinatio**
The following operations are supported:

#### Create

```
$ curl -X POST -d @la_sagne.kml -H "Content-Type: application/vnd.google-earth.kml+xml" -H "Referer: http://map.geo.admin.ch" \
 http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files

{"adminId":"J3KMLz8CRyKjO-n1DhMm7w_admin","fileId":"J3KMLz8CRyKjO-n1DhMm7w"}
```

#### GET with file_id

```
$ curl http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files/J3KMLz8CRyKjO-n1DhMm7w

<?xml version="1.0" encoding="UTF-8"?><kml xmlns="http://www.opengis.net/kml/2.2".....
```

#### GET with admin_id

```
$ curl http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files/J3KMLz8CRyKjO-n1DhMm7w

{"fileId":"J3KMLz8CRyKjO-n1DhMm7w"}
```

#### POST with admin_id (UPDATE)

```
$ curl -v -X POST -d @la_sagne.kml  -H "Content-Type: application/vnd.google-earth.kml+xml" -H "Referer: http://map.geo.admin.ch"  \
 http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files/J3KMLz8CRyKjO-n1DhMm7w_admin

{"adminId":"J3KMLz8CRyKjO-n1DhMm7w_admin","fileId":"J3KMLz8CRyKjO-n1DhMm7w"}
```

#### POST with file_id (FORK)

```
$ curl -v -X POST -d @la_sagne.kml  -H "Content-Type: application/vnd.google-earth.kml+xml" -H "Referer: http://map.geo.admin.ch" \
 http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files/J3KMLz8CRyKjO-n1DhMm7w

{"adminId":"hghdthtfghdh_admin","fileId":"hghdthtfghdh"}
```
#### DELETE with admin_id
```
$ curl -X DELETE  -H "Referer: http://map.geo.admin.ch" http://mf-chsdi3.dev.bgdi.ch/mom_kml_storage/files/J3KMLz8CRyKjO-n1DhMm7w_admin

HTTP/1.1 200 OK
```

Currently, only KML files are supported. Files are check for correct 'Content-Type', size and well-formedness, and are stored into bucket `public.geo.admin.ch` (or the equivalent on stagging environnments)

## TODO

- [x] View for each operations
- [x] Multiple buckets + Varnish settings
- [x] Full HTTPS check (new certificate for _public.geo.admin.ch_!)
- [x] Automatic tests
- [x] Proper credentials management
- [x] Dynamodb dump/restore
- [ ] Bucket backup
- [x] Proper scaling (how much traffic to we expect, how many files ?)

From https://github.com/geoadmin/mf-geoadmin3/issues/2315
- [ ]  availability to manage the data by KOGIS (DB mmgt interfce, instruction) (see https://github.com/geoadmin/dynamodump/pull/15)
- [ ] set up automatic deletion date (bucket policy)

## Backups

* DynamoDB: see https://github.com/geoadmin/dynamodump/pull/15
* S3 Bucket: `s3sync`

## Integration in geo.admin.ch

Test [link](https://mf-geoadmin3.dev.bgdi.ch/teo_save_draw)

### CODE REVIEW

**Comment by procrastinatio** [comment]
OK, jenkins will breaks as long as the boto credentials are not also on jenkins

**Comment by gjn** [comment]
We have boto credentials for dynamodb (shortener). Could we re-use the same?

File: chsdi/views/files.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        if request.matched_route.name == 'files':
+            self.file_id = request.matchdict['id']
+            try:
+                key = self.bucket.get_key(self.file_id)
+                if key is not None:
+                    self.key = key
+                else:
+                    raise exc.HTTPNotFound('File %s not found' % self.file_id)
+            except:
+                raise exc.HTTPInternalError('Cannot access file with id=%s' % self.file_id)
```
Comment (loicgasser) [inline chsdi/views/files.py]: AttributeError: 'module' object has no attribute 'HTTPInternalError'

File: chsdi/views/files.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            self.admin_id = self.file_id + '_admin'
+            mime = self.request.content_type
+            data = self.request.body
+            try:
+                k = Key(self.bucket)
+                k.set_metadata('Content-Type', mime)
+                k.set_contents_from_string(data, replace=False)
+                return {'adminId': self.admin_id, 'fileId': self.file_id}
+            except:
+                raise exc.HTTPInternalError('Cannot create file')
```
Comment (loicgasser) [inline chsdi/views/files.py]: HTTPInternalServerError instead

File: chsdi/views/files.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        self.admin_id = self.file_id + '_admin'
+        mime = self.request.content_type
+        data = self.request.body
+        try:
+            k = Key(self.bucket)
+            k.set_metadata('Content-Type', mime)
+            k.set_contents_from_string(data, replace=False)
+            return {'adminId': self.admin_id, 'fileId': self.file_id}
+        except:
+            raise exc.HTTPInternalError('Cannot create file')
```
Comment (loicgasser) [inline chsdi/views/files.py]: HTTPInternalServerError

File: chsdi/views/files.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    @requires_authorization()
+    @validate_kml_input()
+    def update_file(self):
+        data = self.request.body
+        if self.admin_id is not None:
+            try:
+                self.key.set_contents_from_string(data, replace=True)
+                return {'adminId': self.admin_id, 'fileId': self.file_id}
+            except:
+                raise exc.HTTPInternalError('Cannot save file')
```
Comment (loicgasser) [inline chsdi/views/files.py]: HTTPInternalServerError

**Comment by oterral** [comment]
jenkins test

**Comment by procrastinatio** [comment]
@gjn No, I think it is better to define a new user for the table and s3 bucket used by this service.

File: chsdi/views/files.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    table = _get_dynamodb_table()
+    try:
+        table.get_item(adminId=str(adminId))
+    except ItemNotFound:
+        return False
+
+    return True
+
+
+def _get_file_id_from_admin_id(admin_id):
```
Comment (loicgasser) [inline chsdi/views/files.py]: in order to be consitent use `adminId` here as well.

File: chsdi/views/files.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    # TODO: get the file id from database
+    fileId = None
+    table = _get_dynamodb_table()
+    try:
+        item = table.get_item(adminId=str(admin_id))
+        fileId = item.get('fileId')
+    except Exception as e:
+        raise exc.HTTPBadRequest('The id %s doesn\'t exist. Error is: %s' % (id, e))
+    if fileId is None:
```
Comment (loicgasser) [inline chsdi/views/files.py]: when is there an `adminId` but no `fileId` ?

File: chsdi/views/files.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    return fileId
+
+
+@view_defaults(renderer='jsonp', route_name='files')
+class FileView(object):
+
+    def __init__(self, request):
+        self.request = request
+        self.bucket = self._get_bucket()
+        if request.matched_route.name == 'files':
```
Comment (loicgasser) [inline chsdi/views/files.py]: or you sure you need that? It is already defined in `view_defaults` above.

File: chsdi/views/files.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            self.admin_id = None
+            id = request.matchdict['id']
+            if _is_admin_id(id):
+                self.admin_id = id
+                self.file_id = _get_file_id_from_admin_id(self.admin_id)
```
Comment (loicgasser) [inline chsdi/views/files.py]: minor again: admin_id vs adminId

File: chsdi/views/files.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            else:
+                self.file_id = id
+            try:
+                key = self.bucket.get_key(self.file_id)
+                if key is not None:
+                    self.key = key
+                else:
+                    raise exc.HTTPNotFound('File %s not found' % self.file_id)
+            except:
+                raise exc.HTTPInternalServerError('Cannot access file with id=%s' % self.file_id)
```
Comment (loicgasser) [inline chsdi/views/files.py]: here you'll always raise `HTTPInternalServerError` , if you want to raise `HTTPNotFound` I believe you need to put it outside the try/except

File: chsdi/views/files.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+    def _get_uuid(self):
+        return base64.urlsafe_b64encode(uuid.uuid4().bytes).replace('=', '')
+
+    def _get_bucket(self):
+        # TODO use profile instead when correctly installed
+        PROFILE_NAME = 'geoadmin-filestorage'
```
Comment (loicgasser) [inline chsdi/views/files.py]: to put in settings as well

File: chsdi/views/files.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        config.read(["/home/ltmom/.boto"])
+
+        access_key = config.get(PROFILE_NAME, 'aws_access_key_id')
+        secret_key = config.get(PROFILE_NAME, 'aws_secret_access_key')
+        try:
+            # TODO use this
+            #conn = S3Connection(profile_name='wmts_test')
+            conn = S3Connection(aws_access_key_id=access_key, aws_secret_access_key=secret_key)
+            bucket = conn.get_bucket(BUCKET_NAME)
+        except Exception as e:
```
Comment (loicgasser) [inline chsdi/views/files.py]: to be more accurate.
If the bucket does not exist, an S3ResponseError will be raised.

**Comment by procrastinatio** [comment]
jenkins test

File: chsdi/views/files.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        except:
+            raise exc.HTTPInternalServerError('Cannot create file')
+
+    @view_config(request_method='GET')
+    def read_file(self):
+        try:
+            if self.admin_id is not None:
+                return {'fileId': self.file_id}
+            data = self.key.get_contents_as_string()
+            response = Response(data, content_type=self.key.content_type)
```
Comment (loicgasser) [inline chsdi/views/files.py]: minor: `return Response` one line less :+1:

File: chsdi/views/files.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    def create_file(self):
+        self.file_id = self._get_uuid()
+        self.admin_id = self._get_uuid()
+        mime = self.request.content_type
+        data = self.request.body
+        try:
+            k = Key(self.bucket)
+            k.key = self.file_id
+            k.set_metadata('Content-Type', mime)
+            k.set_contents_from_string(data, replace=False)
```
Comment (loicgasser) [inline chsdi/views/files.py]: gzipped?

File: chsdi/views/files.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+            try:
+                k = Key(self.bucket)
+                k.key = self.file_id
+                k.set_metadata('Content-Type', mime)
+                k.set_contents_from_string(data, replace=False)
+                _add_item(self.admin_id, self.file_id)
+
+                return {'adminId': self.admin_id, 'fileId': self.file_id, 'status': 'copied'}
+            except:
+                raise exc.HTTPInternalServerError('Cannot create file')
```
Comment (loicgasser) [inline chsdi/views/files.py]: would be nice to merge it with `files_collection` as this is duplicated code.

File: chsdi/views/files.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+    @view_config(request_method='DELETE')
+    @requires_authorization()
+    def delete_file(self):
+        try:
+            if self.admin_id is not None:
+                self.bucket.delete_key(self.key)
```
Comment (loicgasser) [inline chsdi/views/files.py]: You need to delete the entry in dynamodb as well no?

File: chsdi/views/files.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+                return Response(body={'success': True}, status='200 OK')
+            else:
+                raise exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)
+        except:
+            raise exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)
+
+    @view_config(request_method='OPTIONS', renderer='string')
+    def options_file(self):
```
Comment (loicgasser) [inline chsdi/views/files.py]: I am not sure I understand very well the point of this view. Could you explain it to me?

File: chsdi/views/files.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+    @requires_authorization()
+    def delete_file(self):
+        try:
+            if self.admin_id is not None:
+                self.bucket.delete_key(self.key)
+                return Response(body={'success': True}, status='200 OK')
+            else:
+                raise exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)
+        except:
+            raise exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)
```
Comment (loicgasser) [inline chsdi/views/files.py]: I would rewrite that like this:

    if self.admin_id is None:
        raise exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)
    else:
        try:
            self.bucket.delete_key(self.key)
            Response(body={'success': True}, status='200 OK')
        except:
             exc.HTTPInternalServerError('Cannot delete file %s' % self.file_id)

File: chsdi/views/files.py [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+
+            return func(self, *args, **kwargs)
+        return wrapper
+    return decorator
+
+
+def _get_dynamodb_table():
+    PROFILE_NAME = 'Credentials'
+    config = ConfigParser.ConfigParser()
+    config.read(["/etc/boto.cfg"])
```
Comment (loicgasser) [inline chsdi/views/files.py]: This should change ultimately right?

**Comment by loicgasser** [comment]
:+1:  Nice, a couple of comments on a stil WIP PR, a few things to address, but I think it looks pretty good.

File: chsdi/views/files.py (hunk 6 above)
Comment (procrastinatio) [inline chsdi/views/files.py]: Should never be the case. You only have *one* id, but you only dynamoDB knows what is an *fileId* and what is an *adminId*.

File: chsdi/views/files.py [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+
+    def _get_uuid(self):
+        return base64.urlsafe_b64encode(uuid.uuid4().bytes).replace('=', '')
+
+    def _get_bucket(self):
+        # TODO use profile instead when correctly installed
+        PROFILE_NAME = 'geoadmin-filestorage'
+        BUCKET_NAME = self.request.registry.settings['geoadmin_file_storage_bucket']
+        config = ConfigParser.ConfigParser()
+        config.read(["/home/ltmom/.boto"])
```
Comment (gjn) [inline chsdi/views/files.py]: This shouldn't end up in master...

File: chsdi/views/files.py (hunk 14 above)
Comment (gjn) [inline chsdi/views/files.py]: +1

**Comment by gjn** [comment]
This looks good. I think a couple of Loic's and my comments should be addressed before merging it.

File: chsdi/views/files.py (hunk 13 above)
Comment (procrastinatio) [inline chsdi/views/files.py]: No, it was decided, we do not gzip.

File: chsdi/views/files.py (hunk 8 above)
Comment (procrastinatio) [inline chsdi/views/files.py]: OK

File: chsdi/views/files.py (hunk 5 above)
Comment (procrastinatio) [inline chsdi/views/files.py]: OK

File: chsdi/views/files.py (hunk 13 above)
Comment (loicgasser) [inline chsdi/views/files.py]: why not?

Comment (cedricmoullet) [inline chsdi/views/files.py]: The idea was not to store KMZ, but to send gzipped content. Is it doable ?

**Comment by cedricmoullet** [comment]
@procrastinatio please squash and merge

**Comment by gjn** [comment]
@procrastinatio Do you agree to merge this and deploy it today?

**Comment by procrastinatio** [comment]
Yes, you may merge.

**Pull request merged**